### PR TITLE
Refactor e-mail verification to use VerificationCode model

### PR DESCRIPTION
**Pull request by mr-kojo**
Based on issue #275.

I'll continue right after the Christmas holidays 🎄

### CODE REVIEW

**Comment by mr-kojo** [comment]
Meanwhile, I discovered [`django.contrib.auth.tokens.PasswordResetTokenGenerator`](https://github.com/django/django/blob/master/django/contrib/auth/tokens.py). This class creates and checks **non-persistent** password reset verification codes. Here, a verification code is simply a hash of the user object together with the current timestamp. Thus, when the user opens the password reset URL and enters a new password, the verification code gets invalidated automatically (because the user object changes and thus the hash changes, too). 
This mechanism could be used for email verification, email change, and account deletion just as well.

Now I wonder if I should implement it this way (or if I should stick to the solution with persistent verification codes).

PROS:
- Time-saving: Re-using code, no need to implement and maintain verification code functionality
- Non-persistent: No need to cram the database with (single-usage) verification codes
- Secure: Well-tested Django implementation using cryptographic hashes with salt

CONS:
- Less flexibility: This approach requires that the user object must not be modified by the time the user opens the verification URL. Hence, the user must not modify his/her profile in the meantime and she/he must not login again (because this would modify the `last_login` field). However, all of this could be enforced quite easily: We could simply log out the user and disable login until he/she opens the verification URL. 
- Less robust: If a user requests a verification mail and right afterwards an admin modifies the corresponding user for some reason, this will break the verification code. In such a case, the user will have to request a new verification mail.
- Harder to debug: Since verification codes are not stored persistently, there is no way to see e.g. if a user has requested one already and if so, if/when he/she has used it.

What are your opinions?

**Comment by tiltec** [comment]
Nice find! I like solutions where we can avoid storing tokens in the database.

I think we can use the `PasswordResetTokenGenerator` in two different ways: for password reset directly and for other user-related tokens when subclassed.

In the subclass, I would modify theses bits:
- rename `settings.PASSWORD_RESET_TIMEOUT_DAYS` to something else to make it clear that it's not only used for password reset: https://github.com/django/django/blob/master/django/contrib/auth/tokens.py#L49
- remove dependency on user.last_login in `_make_hash_value()` 

Maybe something like this:
```
class TokenGenerator(PasswordResetTokenGenerator):
    def _make_hash_value(self, user, timestamp):
        return (str(user.pk) + user.password + str(timestamp))
```
[source](https://medium.com/@frfahim/django-registration-with-confirmation-email-bb5da011e4ef)

I'm unsure if we should also use a similar token generator for invitation codes, as I can't imagine a reliable source to calculate the hash from.

**Comment by tiltec** [comment]
I had a look at [django-registration](https://django-registration.readthedocs.io/en/2.3/hmac.html), it calculates the hash only from the username field, adds a configurable salt to it and signs with Djangos `SECRET_KEY`:
https://github.com/ubernostrum/django-registration/blob/master/registration/backends/hmac/views.py#L64-L66

**Comment by mr-kojo** [comment]
> I had a look at django-registration, it calculates the hash only from the username field, adds a configurable salt to it and signs with Djangos SECRET_KEY

Great, even more elegant! That's how I'll do it 👍

File: AUTHORS.md [hunk 1]
```diff
@@ -3,6 +3,7 @@
 - [Douglas Webb](https://github.com/bbew/), dmhwebb@gmail.com
 - [Florian Gössele](https://github.com/goessek/), goessek@posteo.de
 - [Kristijan Miklobusec](https://github.com/Sliverriver), kristijan80@web.de
+- [Lukas Gebhard](https://github.com/mr-kojo), freerunningapps@gmail.com
```
Comment (tiltec) [inline AUTHORS.md]: Oh the authors file! I think you're one of the few people listed here that contributed code to karrot ;)

Comment (mr-kojo) [reply AUTHORS.md]: In particular, @nicksellen is missing

Comment (tiltec) [reply AUTHORS.md]: Yes, I wonder why we need this file at all. Git history is much more useful.

**Comment by mr-kojo** [comment]
I found two problems of my current implementation:

1)  A user can use the same verification code twice e.g. if he/she changes his/her email address twice within the time limit that is configured for email verification.
2) If a user changes his/her password and email at once and then opens the email verification URL, this will break the verification code for password change (because a verification code contains a hash of the email address).

I have been thinking about this for quite some time now and the best solution I found is this:

To solve (1):
- Before a verification code is generated, we ensure all previous ones have to be validated (used) already.
- Each verification code has an index number. The most recent index is stored in the database. When validating a verification code, we check if the index is lower than the one in the database. If so, the verification code is invalid because it has already been used before.

To solve (2):
We hash the user's primary key instead of the email address.

I am currently implementing this...
Let me know if you have concerns, though.

**Comment by nicksellen** [comment]
I wonder if it's simpler to just put them in the database, explicit over implicit. I think it's easy to run into caveats and funny edge cases when doing clever things.

**Comment by tiltec** [comment]
The fix for (1) looks like it would defeat the advantage of not having to store process-related data in the database. Maybe then it's better to save the code directly in the DB or what would speak against it?

**Comment by mr-kojo** [comment]
You're right. Too bad, would've liked the non-persistent approach. 
Anyway, it was an interesting detour :)
Lesson learnt: Non-persistent verification codes are a clever way to authenticate actions that
- can only be done once and can't be undone (e.g., activate a non-editable username), or
- have to be done immediately (i.e., don't let the user do anything else before she/he has performed the action).

**Comment by tiltec** [comment]
It's an interesting concept and it still seems suitable for activating the email/account. But now that we go for an explicit code implementation, maybe we'll also do email activation with it.

**Comment by mr-kojo** [comment]
Ready to be merged (at least I hope so)

**Comment by tiltec** [comment]
Thanks! I got the notification, will review tomorrow.
Is it compatible to the current frontend version?

File: foodsaving/users/migrations/0019_user_photo.py [hunk 2]
```diff
@@ -7,13 +7,14 @@
 class Migration(migrations.Migration):
 
     dependencies = [
-        ('users', '0017_auto_20171207_1210'),
+        ('users', '0018_remove_verification_codes'),
```
Comment (tiltec) [inline foodsaving/users/migrations/0019_user_photo.py]: I think you must put new migrations after old migrations that already ran. I fear some problems on staging and master databases when inserting a new migrations between two old ones.

File: foodsaving/users/models.py [hunk 3]
```diff
@@ -149,9 +141,9 @@ def _send_mail_change_notification(self):
     def _send_welcome_mail(self):
         self._unverify_mail()
 
-        url = '{hostname}/#/verify-mail?key={key}'.format(
+        url = '{hostname}/#/verify-mail?code={code}'.format(
```
Comment (tiltec) [inline foodsaving/users/models.py]: Ah, the rename from `key` to `code` will break the frontend. Would you rather change the frontend or change it back to `key={code}`?

Also, we could also use the `furl` library to construct the URL (similar to the invitation mail). But not necessary...

File: foodsaving/users/models.py [hunk 4]
```diff
@@ -172,9 +164,9 @@ def _send_welcome_mail(self):
     def send_new_verification_code(self):
         self._unverify_mail()
 
-        url = '{hostname}/#/verify-mail?key={key}'.format(
+        url = '{hostname}/#/verify-mail?code={code}'.format(
```
Comment (tiltec) [inline foodsaving/users/models.py]: Same here about frontend break.

File: foodsaving/userauth/serializers.py [hunk 5]
```diff
@@ -41,7 +42,7 @@ class Meta:
         model = get_user_model()
         fields = ['id', 'display_name', 'email', 'unverified_email', 'password',
                   'address', 'latitude', 'longitude', 'description', 'mail_verified',
-                  'key_expires_at', 'current_group', 'language', 'photo', 'photo_urls']
+                  'current_group', 'language', 'photo', 'photo_urls']
         read_only_fields = ('unverified_email', 'key_expires_at', 'mail_verified')
```
Comment (tiltec) [inline foodsaving/userauth/serializers.py]: Remove `key_expires_at` from `read_only_fields` too

File: foodsaving/users/migrations/0017_auto_20171207_1210.py [hunk 6]
```diff
@@ -16,6 +16,7 @@ class Migration(migrations.Migration):
         migrations.AlterField(
             model_name='user',
             name='current_group',
-            field=models.ForeignKey(blank=True, null=True, on_delete=django.db.models.deletion.SET_NULL, to='groups.Group'),
+            field=models.ForeignKey(blank=True, null=True, on_delete=django.db.models.deletion.SET_NULL,
```
Comment (tiltec) [inline foodsaving/users/migrations/0017_auto_20171207_1210.py]: unrelated change

File: foodsaving/groups/migrations/0017_auto_20171207_1210.py [hunk 7]
```diff
@@ -16,6 +16,7 @@ class Migration(migrations.Migration):
         migrations.AlterField(
             model_name='group',
             name='active_agreement',
-            field=models.OneToOneField(null=True, on_delete=django.db.models.deletion.SET_NULL, related_name='active_group', to='groups.Agreement'),
+            field=models.OneToOneField(null=True, on_delete=django.db.models.deletion.SET_NULL,
```
Comment (tiltec) [inline foodsaving/groups/migrations/0017_auto_20171207_1210.py]: Unrelated change, too

File: foodsaving/userauth/api.py [hunk 8]
```diff
@@ -100,7 +98,7 @@ def post(self, request):
         """
         Send token to verify e-mail
 
-        requires "key" parameter
+        requires "code" parameter
```
Comment (tiltec) [inline foodsaving/userauth/api.py]: Serializer still expects the `key` parameter, I would probably leave it that way for now.

File: foodsaving/userauth/models.py [hunk 9]
```diff
@@ -0,0 +1,57 @@
+from datetime import timedelta
+
+from django.db.models import CharField, ForeignKey, Manager
+from django.db import models
+from django.conf import settings
+from django.utils import crypto, timezone
+
+from foodsaving.base.base_models import BaseModel
+
+# TODO: Ensure each user can have at most one verification code of a certain type at a time.
```
Comment (tiltec) [inline foodsaving/userauth/models.py]: Seems good, you can do that with a one-liner by adding a `unique_together` constraint. I used it in the invitations code, for example.

File: foodsaving/users/models.py (hunk 3 above)
Comment (mr-kojo) [reply foodsaving/users/models.py]: Oh... sorry for that!
I actually wanted to test everything with my local frontend, then I would have noted that myself. But then I forgot...

**Comment by mr-kojo** [comment]
Successfully tested using my local frontend (should have done that right away, sorry for that)

File: foodsaving/conversations/migrations/0007_auto_20171207_1210.py [hunk 10]
```diff
@@ -16,6 +16,7 @@ class Migration(migrations.Migration):
         migrations.AlterField(
             model_name='conversationparticipant',
             name='seen_up_to',
-            field=models.ForeignKey(null=True, on_delete=django.db.models.deletion.SET_NULL, to='conversations.ConversationMessage'),
+            field=models.ForeignKey(null=True, on_delete=django.db.models.deletion.SET_NULL,
```
[the remark this answers is not in the document]
Comment (tiltec) [reply foodsaving/conversations/migrations/0007_auto_20171207_1210.py]: Yes, migrations are important! On the other hand, they have been autogenerated in 99% of the cases and the Django migrations generator didn't respect PEP8 out of the box (I didn't try further), so I just said that PEP8 wouldn't be checked there.
Another point is that migrations are usually "throw away code": after they ran successfully, they can be [squashed](https://docs.djangoproject.com/en/2.0/topics/migrations/#migration-squashing) and removed. Edits should _never_ be necessary after they've been merged to master.

Before we start enforcing PEP8 on migrations, I would strongly like to have a `makemigrations` command that produces compliant files. Django won't do it out of the box, see [this bug](https://code.djangoproject.com/ticket/24275), which has been closed as WONTFIX. The solution could be to override the `makemigrations` to run something like `autopep8` automatically.

You are fine to write your custom migrations in PEP8, but I consider style changes to old migrations unnecessary noise (which could have been avoided by having all files in PEP8 from day one)

**Comment by tiltec** [comment]
Cool, looks even better now! 

If you still want to revert the unrelated changes, I'd be very happy. I'm not strongly bothered by it, but would say that it's easier to review a PR which contains only changes related to a single issue.
